packages/app/src/vis-packs/core/compound/utils.ts

```diff
@@ -1,14 +1,21 @@
+import type { Notation } from '@h5web/lib';
 import type { Primitive, PrintableType } from '@h5web/shared';
-import { formatMatrixValue } from '@h5web/shared';
-import { formatMatrixComplex } from '@h5web/shared';
 
-export function compoundFormatter(value: Primitive<PrintableType>) {
+import {
+  createMatrixComplexFormatter,
+  createNumericFormatter,
+} from '../matrix/utils';
+
+export function compoundFormatter(
+  value: Primitive<PrintableType>,
+  notation: Notation
+) {
   if (Array.isArray(value)) {
-    return formatMatrixComplex(value);
+    return createMatrixComplexFormatter(notation)(value);
   }
 
   if (typeof value === 'number') {
-    return formatMatrixValue(value);
+    return createNumericFormatter(notation)(value);
```

Comment:
I can't use the updated `getFormatter` as I have access to the `value` here and not to the `dataset`.

Reply:
You're recreating a formatter for every `value` here, though. You should write a `createCompoundFormatter(notation)` function, initialise all the formatters you may need once, and then return a formatter arrow function that uses all these formatters.

That being said, if I remember correctly, the source of the problem is that the `formatter` only receives a `value`, and since we're dealing with a compound dataset, we can't know what the type of that value will be ahead of time.

Could `MatrixVis` maybe pass the `col` index to the `formatter`? Wouldn't this be enough information for `MappedCompoundMatrixVis` to find the correct sub-type for that column? You could then refactor `getFormatter` to accept a `DType` rather than a `Dataset`.